module/executiondatasync/requester/dispatcher.go

```diff
@@ -0,0 +1,264 @@
+package requester
+
+import (
+	"context"
+	"fmt"
+	"sort"
+
+	"github.com/rs/zerolog"
+
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/util"
+	"github.com/onflow/flow-go/storage"
+)
+
+type resultInfo struct {
+	resultID        flow.Identifier
+	blockID         flow.Identifier
+	blockHeight     uint64
+	executionDataID flow.Identifier
+	sealed          bool
+}
+
+type dispatcher struct {
+	sealedHeight uint64
+	jobs         map[uint64]map[flow.Identifier]context.CancelFunc
+
+	receiptsIn         chan<- interface{}
+	receiptsOut        <-chan interface{}
+	finalizedBlocksIn  chan<- interface{}
+	finalizedBlocksOut <-chan interface{}
+	resultInfosIn      chan<- interface{}
+	resultInfosOut     <-chan interface{}
+
+	blocks  storage.Blocks
+	results storage.ExecutionResults
+
+	handler   *handler
+	fulfiller *fulfiller
+
+	logger  zerolog.Logger
+	metrics module.ExecutionDataRequesterMetrics
+
+	component.Component
+}
+
+func newDispatcher(
+	sealedHeight uint64,
+	blocks storage.Blocks,
+	results storage.ExecutionResults,
+	handler *handler,
+	fulfiller *fulfiller,
+	logger zerolog.Logger,
+	metrics module.ExecutionDataRequesterMetrics,
+) *dispatcher {
+	receiptsIn, receiptsOut := util.UnboundedChannel()
+	finalizedBlocksIn, finalizedBlocksOut := util.UnboundedChannel()
+	resultInfosIn, resultInfosOut := util.UnboundedChannel()
```

Comment:
as I mentioned above, the reason I initially introduced them is because they allow the rate of production and consumption to be independent of each other. This is better in situations where overall the consumer is able to keep up with the rate of consumption, but there may be moments during which the consumer is processing items and producers would otherwise block during those moments if we didn't use an unbuffered channel

Reply:
Let me know what you think about this, I'm not sure which approach is better in terms of tradeoffs.